### PR TITLE
ResolVBUS binding

### PR DESCRIPTION
**Pull request by MitchSUEW**
This is the initial contribution of the Resol VBUS Binding

The WIKI can be found at https://github.com/openhab/openhab/wiki/Resol-VBUS-Binding

### CODE REVIEW

**Comment by teichsta** [comment]
Hi @MitchSUEW,

thanks for this contribution! By scanning your PR i came across many files (.project) which shouldn't belong to this PR. Please make sure that only changes regarding ResolVBUS are contained.

Hope to find the time for an in depth review soon.

Best, Thomas E.-E.

**Comment by MitchSUEW** [comment]
@teichsta Now we have only additions and no deletions. And everything only in the resolvbus namespace.

**Comment by watou** [comment]
In the wiki page, could you include the contents of the Google Sheets link in the wiki itself and remove the link to Google Sheets?  [Building tables in Github Flavored Markdown](https://help.github.com/articles/github-flavored-markdown/#tables) is pretty easy, and it would help to keep all of the binding's documentation in the one place.

Also, the last line of your openhab.cfg fragment is incorrect.

File: bundles/binding/org.openhab.binding.resolvbus/.classpath [hunk 1]
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<classpath>
+	<classpathentry kind="src" output="target/classes" path="src/main/java"/>
+	<classpathentry excluding="**" kind="src" output="target/classes" path="src/main/resources"/>
+	<classpathentry kind="con" path="org.eclipse.jdt.launching.JRE_CONTAINER/org.eclipse.jdt.internal.debug.ui.launcher.StandardVMType/JavaSE-1.7"/>
+	<classpathentry kind="con" path="org.eclipse.pde.core.requiredPlugins"/>
+	<classpathentry kind="lib" path="lib/xstream-1.4.6.jar" sourcepath="/Users/michael/.m2/repository/com/thoughtworks/xstream/xstream/1.4.6/xstream-1.4.6-sources.jar"/>
```
Comment (watou) [inline bundles/binding/org.openhab.binding.resolvbus/.classpath]: Please remove sourcepath

File: bundles/binding/org.openhab.binding.resolvbus/META-INF/MANIFEST.MF [hunk 2]
```diff
@@ -0,0 +1,32 @@
+Manifest-Version: 1.0
+Private-Package: org.openhab.binding.resolvbus.internal
+Ignore-Package: org.openhab.binding.resolvbus.internal
+Bundle-License: http://www.eclipse.org/legal/epl-v10.html
+Bundle-Name: openHAB ResolVBUS Binding
+Bundle-SymbolicName: org.openhab.binding.resolvbus
+Bundle-Vendor: openHAB.org
+Bundle-Version: 1.7.0.qualifier
```
Comment (watou) [inline bundles/binding/org.openhab.binding.resolvbus/META-INF/MANIFEST.MF]: Please change to 1.8.0.qualifier

File: bundles/binding/org.openhab.binding.resolvbus/pom.xml [hunk 3]
```diff
@@ -0,0 +1,35 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+	xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
+
+	<parent>
+		<groupId>org.openhab.bundles</groupId>
+		<artifactId>binding</artifactId>
+		<version>1.7.0-SNAPSHOT</version>
```
Comment (watou) [inline bundles/binding/org.openhab.binding.resolvbus/pom.xml]: Please change to 1.8.0-SNAPSHOT

File: bundles/binding/org.openhab.binding.resolvbus/src/main/java/org/openhab/binding/resolvbus/internal/ResolVBUSBinding.java [hunk 4]
```diff
@@ -0,0 +1,406 @@
...
+	private String serialPort;
+	private int inputMode;
+	private long updateInterval;
+	private String deviceID;
+	private static final int INPUT_MODE_LAN = 10;
+	private static final int INPUT_MODE_SERIAL = 20;
+	private boolean useThread = true;
+	private String deviceIDFile;
+	private String deviceIDFromConfig;
+	private static final String bindingversion = "1.0.0";
```
Comment (watou) [inline bundles/binding/org.openhab.binding.resolvbus/src/main/java/org/openhab/binding/resolvbus/internal/ResolVBUSBinding.java]: Please remove.

File: bundles/binding/org.openhab.binding.resolvbus/src/main/java/org/openhab/binding/resolvbus/internal/ResolVBUSBinding.java [hunk 5]
```diff
@@ -0,0 +1,406 @@
...
+
+	private void stopListener() {
+		logger.debug("Stoppig ResolVBUS listener...");
+		if (packetReceiver != null)
+			packetReceiver.stopListener();
+	}
+	
+	/**
+	 * @{inheritDoc}
+	 */
```
Comment (watou) [inline bundles/binding/org.openhab.binding.resolvbus/src/main/java/org/openhab/binding/resolvbus/internal/ResolVBUSBinding.java:237]: Use {@inheritdoc}

File: bundles/binding/org.openhab.binding.resolvbus/src/main/java/org/openhab/binding/resolvbus/internal/ResolVBUSBinding.java [hunk 6]
```diff
@@ -0,0 +1,406 @@
...
+	@Override
+	protected long getRefreshInterval() {
+		return refreshInterval;
+	}
+
+	/**
+	 * @{inheritDoc}
+	 */
```
Comment (watou) [inline bundles/binding/org.openhab.binding.resolvbus/src/main/java/org/openhab/binding/resolvbus/internal/ResolVBUSBinding.java:245]: Use {@inheritdoc}

File: bundles/binding/org.openhab.binding.resolvbus/src/main/java/org/openhab/binding/resolvbus/internal/ResolVBUSBinding.java [hunk 7]
```diff
@@ -0,0 +1,406 @@
...
+	@Override
+	protected String getName() {
+		return "ResolVBUS Refresh Service";
+	}
+	
+	/**
+	 * @{inheritDoc}
+	 */
```
Comment (watou) [inline bundles/binding/org.openhab.binding.resolvbus/src/main/java/org/openhab/binding/resolvbus/internal/ResolVBUSBinding.java:253]: Use {@inheritdoc}

File: bundles/binding/org.openhab.binding.resolvbus/src/main/java/org/openhab/binding/resolvbus/internal/ResolVBUSBinding.java [hunk 8]
```diff
@@ -0,0 +1,406 @@
...
+	
+	public void loadXMLConfig()  {
+		
+		if (deviceID == null) {
+			logger.info("Device ID was null, reading from openhab.cfg");
+			deviceID = deviceIDFromConfig;
+		}
+		
+		try {
+//			deviceID="7101";
```
Comment (watou) [inline bundles/binding/org.openhab.binding.resolvbus/src/main/java/org/openhab/binding/resolvbus/internal/ResolVBUSBinding.java]: Please remove.

File: bundles/binding/org.openhab.binding.resolvbus/src/main/java/org/openhab/binding/resolvbus/internal/ResolVBUSBinding.java [hunk 9]
```diff
@@ -0,0 +1,406 @@
...
+			logger.debug("Loading XML-Config for device ID: {}",deviceID);
+			logger.info("Using Config for device ID: {}",deviceID);
```
Comment (watou) [inline bundles/binding/org.openhab.binding.resolvbus/src/main/java/org/openhab/binding/resolvbus/internal/ResolVBUSBinding.java]: No need for both log entries; please remove .info

File: bundles/binding/org.openhab.binding.resolvbus/src/main/java/org/openhab/binding/resolvbus/internal/ResolVBUSBinding.java [hunk 10]
```diff
@@ -0,0 +1,406 @@
...
+			
+			URL mapping = FrameworkUtil.getBundle(ResolVBUSBinding.class).getEntry("xml/mapping.cfg");
+			
+			if (mapping == null) {
+				config = null;
+				logger.error("Unable to load mapping.xml");
```
Comment (watou) [inline bundles/binding/org.openhab.binding.resolvbus/src/main/java/org/openhab/binding/resolvbus/internal/ResolVBUSBinding.java]: error message does not agree with file name trying to be loaded?

File: bundles/binding/org.openhab.binding.resolvbus/src/main/java/org/openhab/binding/resolvbus/internal/ResolVBUSBinding.java [hunk 11]
```diff
@@ -0,0 +1,406 @@
...
+			InputStream in = mapping.openStream();
+			
+			Properties prop = new Properties();
+			prop.load(in);
+			deviceIDFile = prop.getProperty(deviceID);
+			
+			logger.debug("Using file: {} for configuration",deviceIDFile);
+		} catch (IOException e1) {
+			// TODO Auto-generated catch block
+			e1.printStackTrace();
```
Comment (watou) [inline bundles/binding/org.openhab.binding.resolvbus/src/main/java/org/openhab/binding/resolvbus/internal/ResolVBUSBinding.java]: Use logger; never write to stdout!

**Comment by MitchSUEW** [comment]
@watou Thanks for reviewing, I changed the source according your comments.
Regarding the documentation may I ask you for an advice?
I can include the google spreadsheet into the wiki. But the sheet is WIP and only the beginning and there will be about 70 tabs (each XML-File will be one tab) with each about 2-5 possible devices, each having about 15 datapoints that can be referenced. That would mean a huge page very hard to read.
So I would rather prefer another way. I kind of dynamic selection process:
Choose your Manufacturer => choose your device => here are your datapoints
Is that possible?
Or do you have any other advice how to manage that? 

I also thought of including that into the binding. So when the binding is loaded the first time, you will see which datapoints can be used in the log file. Is that "best practice"? Or hast that been used by anybody?

What would you suggest?

Thanks,
Michael

**Comment by watou** [comment]
Hi Michael,
I was only able to review part of the PR so far, but thanks very much for incorporating the changes.  I will return to reviewing as soon as possible.

I see what you mean about the ultimate volume of information that will be part of the binding documentation regarding supported devices.  Is it that the spreadsheets are intended to be a more readable version of what is present in the binding's XML files?  If that's the case, I recommend either a link directly from the wiki to the part of the repo where the XML files are kept, or to an HTML page or set of pages generated automatically from that data, instead of manually duplicating their contents in a wiki or spreadsheet which will inevitably lead to disagreement and confusion about what devices are supported.  

By comparison, in ESH/OH2 there will be ways to generate documentation from the large body of supported devices that are described in XML files, and I think that how @cdjackson (author/maintainer of the Z-Wave binding) and others are managing this kind of information is probably going to be the best practical approach.  

This might be a radical idea you aren't interested in, but it could make sense to target your binding instead for Eclipse Smarthome and openHAB 2, and merge together this work with how they describe Things, channels, etc., like Chris is doing with the Z-Wave binding.

**Comment by MitchSUEW** [comment]
Just to document a quick update: Since not all data points of all devices are supported yet, I contacted the manufacturer RESOL. Their head of software development Daniel Wippermann will support us and provide "core" classes that will manage data extraction and conversion from their hardware. The surrounding OH binding logic will then be implemented by myself. So stay tuned.
Will work on the documentation when this is finished. Of course considering it to be "OH2-only".

**Comment by teichsta** [comment]
thanks for the update @MitchSUEW!

**Comment by OlMat** [comment]
Thanks for this contribution.
I started to play with OH and connected my Deltasol ES to OH.
My Deltasol ES have 6 temp sensors, volumeflow sensor and solar sensor
Almost everything works ok, but I can't get the Wärmemenge data out from Resol.
My setup have:
`Number  HeatQuantity	"Collected energy [%d Wh]" 	{resolvbus="Wärmemenge"}`
Same data is shown ok in Resol's own application.

**Comment by MitchSUEW** [comment]
@OlMat Sorry to say, that Wärmemenge is currently not implemented. We're in the process of refactoring the whole binding with new classes from RESOL. But due to lack of time from myself there's rather little progress.
Daniel Hillenbrand did something for OH2 which supports the "Wärmemenge"
https://www.codeworkx.de/dl/openhab/

**Comment by 9037568** [comment]
Could @MitchSUEW comment on the state of this PR? Are you doing additional development that should be included before the PR review continues?

**Pull request closed without merging**